File: MATLAB/policyA.m
function f=policyA(aX)

global alpha del beta Nj L theta tau ss0 meaJ


a(1)=0; % INITIAL ASSET 
a(Nj+1)=0;

a(2:Nj)=aX;

K=0;
for j=1:Nj
   K=K+a(j)*meaJ(j);
end

r=alpha*(K/L)^(alpha-1)-del; % INTEREST RATE
w=(1-alpha)*(K/L)^alpha; % WAGE

ss=ss0*w;

% EULER EQUATION (u(c)=log(c))
for j=1:Nj-1
   ff=beta*(1+r)*((1+r)*a(j)+w*theta(j)*(1-tau)+ss(j)-a(j+1)); % beta*(1+r)*c, where c=(1+r)a+w*theta*(1-tau)-a' 
   f(j)=(1+r)*a(j+1)+w*theta(j+1)*(1-tau)+ss(j+1)-a(j+2)-ff;   % 0=c'-ff
end

File: Julia/olg_deterministic/policyA.jl
function policyA(m,aX)
    """
    --------------------------------------------------------------
    === Returns Euler Equation Error given HH's Asset Profile ===
    --------------------------------------------------------------
    ※THIS SHOULD BE ZERO AT EQUILIBRIUM
    <input>
    ・m: model structure 
    ・aX: HH's asset profile(from age 2 to Nj -> (Nj-2+1) element vector)
    <output>
    ・f: Euler equation errors from age 1 to Nj-1 -> (Nj-1+1) element vector
    """

    a = zeros(m.Nj+1);
    # initial asset
    a[1] = 0.0;
    a[m.Nj+1] = 0.0;

    a[2:m.Nj] .= copy(aX);

    K = sum(a[1:end-1] .* m.meaJ); 

    r = m.alpha * (K/m.L)^(m.alpha-1) - m.del; # interest rate
    w = (1-m.alpha)*(K/m.L)^m.alpha; # wage

    ss = m.ss0.*w;

    # euler equation (u(c)=log(c))
    f = zeros(m.Nj-1)
    for j in 1:m.Nj-1
        # beta*(1+r)*c
        ff =  m.beta * (1+r) * ((1+r)*a[j] + w*m.theta[j]*(1-m.tau) + ss[j] - a[j+1]);
        # 0=c'-ff
        f[j] = (1+r)*a[j+1] + w*theta[j+1]*(1-m.tau) + ss[j+1] - a[j+2] - ff;
    end

    return f
end